File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n=s.length();
        int left=0;
        int right=0;
        HashSet<Character> hs= new HashSet<>();
        int maxlength=0;
        while(right<n){
            char c=s.charAt(right);
            while(hs.contains(c)){
                hs.remove(s.charAt(left));
                left++;
            }
            hs.add(c);
            maxlength=Math.max(maxlength, right-left+1);
            right++;

        
    }
    return maxlength;
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        hs= set()
        left=0
        res=0
        for right, ch in enumerate(s):
            while ch in hs:
                hs.remove(s[left])
                left+=1
            res= max(res, right-left+1)
            hs.add(ch)
        return res
            

